oak/server/rust/oak_runtime/src/node/grpc/client.rs

```diff
@@ -63,67 +67,134 @@ impl GrpcClientNode {
         check_uri(&uri)?;
         Ok(Self {
             node_name: node_name.to_string(),
-            handler: GrpcRequestHandler::new(uri, root_tls_certificate),
+            uri,
+            root_tls_certificate,
         })
     }
 
     /// Main loop that handles gRPC invocations from the `handle`, sends gRPC requests to an
     /// external gRPC service and writes gRPC responses back to the invocation channel.
-    async fn handle_loop(
-        &mut self,
-        runtime: RuntimeProxy,
-        handle: Handle,
-    ) -> Result<(), OakStatus> {
+    async fn handle_loop(&self, runtime: RuntimeProxy, handle: Handle) -> Result<(), OakStatus> {
         // Create a [`Receiver`] used for reading gRPC invocations.
         let receiver = Receiver::<Invocation>::new(handle);
         loop {
-            debug!("Waiting for gRPC invocations");
+            debug!("Waiting for gRPC invocation");
             // Read a gRPC invocation from the [`Receiver`].
             let invocation = receiver.receive(&runtime).map_err(|error| {
                 error!("Couldn't receive the invocation: {:?}", error);
                 error
             })?;
 
-            // Receive a request from the invocation channel.
-            let request = invocation.receive_request(&runtime).map_err(|error| {
-                error!(
-                    "Couldn't receive gRPC request from the invocation: {:?}",
-                    error
-                );
+            let result = self.process_invocation(&runtime, &invocation).await;
+            info!("Invocation processing finished: {:?}", result);
+            invocation.close(&runtime);
+        }
+    }
+
+    /// Process a gRPC method invocation for an external gRPC service.
+    async fn process_invocation(
+        &self,
+        runtime: &RuntimeProxy,
+        invocation: &Invocation,
+    ) -> Result<(), OakStatus> {
+        // Receive a request from the invocation channel.
+        let request = invocation.receive_request(&runtime).map_err(|error| {
+            invocation.send_error(rpc::Code::Internal, "Failed to read request", runtime);
+            error!(
+                "Couldn't read gRPC request from the invocation: {:?}",
                 error
+            );
+            error
+        })?;
+        debug!("Incoming gRPC request: {:?}", request);
+
+        // Connect to an external gRPC service.
+        let mut grpc_client = self.connect().await.map_err(|error| {
```

Comment:
Does it mean that a gRPC client will create a new connection for each invocation?

Reply:
Yes, but I think that's also how it behaved before (every invocation of `GrpcRequestHandler::unary_request()` would call `GrpcRequestHandler::connect()` which would overwrite `self.dispatcher`).

I did have a quick attempt at making the connection persistent between method invocations but it didn't immediately work, so I'll add an issue or TODO to investigate further.